src/saveload/marker.rs

```diff
@@ -161,63 +205,73 @@ pub trait Marker: Component + DeserializeOwned + Serialize + Copy {
-    /// Stores mapping `Marker` -> `Entity`.
-    /// If _id_ argument is `Some(id)` then the new marker will have this `id`.
-    /// Otherwise allocator creates marker with a new unique id.
+    /// Allocates a new marker for a given entity.
+    /// If you don't pass an id, a new unique id will be created.
     fn allocate(&mut self, entity: Entity, id: Option<M::Identifier>) -> M;
 
-    /// Get `Entity` by `Marker::Identifier`
-    fn get(&self, id: M::Identifier) -> Option<Entity>;
-
-    /// Create new unique marker `M` and attach it to entity.
-    /// Or get old marker if this entity is already marked.
-    fn mark<'a>(&mut self, entity: Entity, storage: &mut WriteStorage<'a, M>) -> (M, bool) {
-        match storage.get(entity).cloned() {
-            Some(marker) => (marker, false),
-            None => {
-                let marker = self.allocate(entity, None);
-                storage.insert(entity, marker);
-                (marker, true)
-            }
-        }
-    }
+    /// Get an `Entity` by a marker identifier.
+    /// This function only accepts an id; it does not update the marker data.
+    ///
+    /// Implementors usually maintain a marker -> entity mapping
+    /// and use that to retrieve the entity.
+    fn retrieve_entity_internal(&self, id: M::Identifier) -> Option<Entity>;
 
-    /// Find an `Entity` by a `Marker` with same id
-    /// and update `Marker` attached to the instance.
-    /// Or create new entity and mark it.
-    fn get_marked<'a>(
+    /// Tries to retrieve an entity by the id of the marker;
+    /// if no entity has a marker with the same id, a new entity
+    /// will be created and `marker` will be inserted for it.
+    ///
+    /// In case the entity existed,
+    /// this method will update the marker data using `Marker::update`.
+    fn retrieve_entity(
         &mut self,
-        id: M::Identifier,
-        entities: &Entities<'a>,
-        storage: &mut WriteStorage<'a, M>,
+        marker: M,
+        storage: &mut WriteStorage<M>,
+        entities: &EntitiesRes,
     ) -> Entity {
-        if let Some(entity) = self.get(id) {
-            if entities.is_alive(entity) {
+        if let Some(entity) = self.retrieve_entity_internal(marker.id()) {
+            if let Some(marker_comp) = storage.get_mut(entity) {
+                marker_comp.update(marker);
+
                 return entity;
             }
         }
 
         let entity = entities.create();
-        let marker = self.allocate(entity, Some(id));
+        let marker = self.allocate(entity, Some(marker.id()));
+
         storage.insert(entity, marker);
         entity
     }
 
+    /// Create new unique marker `M` and attach it to entity.
+    /// Or get old marker if this entity is already marked.
+    fn mark<'m>(&mut self, entity: Entity, storage: &'m mut WriteStorage<M>) -> (&'m M, bool) {
+        let mut new = false;
+
+        let marker = storage.entry(entity).unwrap().or_insert_with(|| {
```

Comment:
What is with this `unwrap` here? Is it expected to not panic? I would change it to `expect` that contains explanation.